File: plugins/push/ios/PushManager.swift
//
//  PushManager.swift — lecodes-plugins/push
//
//  The device-level push machinery (docs/push-plan.md): the APNs token, the
//  UNUserNotificationCenter moments (via LeCodesAppHooks — see PushPlugin.swift), the
//  le.codes registration calls, and tap/foreground routing. PushService is the thin
//  per-session adapter over it.
//
//  Registration attribution reads engine.currentProjectUuid (the trusted world identity —
//  see LeCodesEngine "World identity & trust"); worlds without an identity (launcher,
//  dev bundles, generated shells) get "unavailable".
//
//  Wire payload (APNs userInfo): { aps: { alert: { title, body? }, badge?, sound },
//  lecodes: { projectUuid, url?, data? } } — `lecodes.projectUuid` is stamped by the
//  backend and routes multi-tenant display/taps; it is stripped before payloads reach JS.
//

import Foundation
import UIKit
import UserNotifications
import LeCodesSDK

public final class PushManager: LeCodesNotificationHandler {

    public static let shared = PushManager()
    private init() {}

    public weak var engine: LeCodesEngine?
    /// The le.codes backend the viewer registers against.
    public var apiBase = "https://le.codes"
    /// APNs gateway of the tokens this build mints — "sandbox" or "production".
    ///
    /// Derived from the `aps-environment` this build was SIGNED with, because that entitlement is
    /// the only thing deciding which APNs environment issues a device token — and therefore which
    /// gateway will accept it. The build CONFIGURATION is not that signal: a Release build on a
    /// development profile still mints sandbox tokens, so the old `#if DEBUG` split reported
    /// "production" for it and every send came back `BadDeviceToken`. Settable, so a host
    /// embedding the SDK can still override.
    public lazy var environment: String = Self.detectApnsEnvironment()

    /// `Entitlements → aps-environment` out of the embedded provisioning profile. The file is CMS
    /// signed, so the payload plist is sliced from the raw bytes by its XML delimiters — a string
    /// round-trip would need a lossy encoding to survive the binary wrapper. App Store builds ship
    /// no embedded profile, and neither does the simulator: production is correct for the former
    /// and the safe default for anything unreadable.
    private static func detectApnsEnvironment() -> String {
        guard let url = Bundle.main.url(forResource: "embedded", withExtension: "mobileprovision"),
              let data = try? Data(contentsOf: url),
              let start = data.range(of: Data("<plist".utf8)),
              let end = data.range(of: Data("</plist>".utf8)),
              start.lowerBound < end.upperBound,
              let root = (try? PropertyListSerialization.propertyList(
                  from: Data(data[start.lowerBound ..< end.upperBound]), format: nil)) as? [String: Any],
              let entitlements = root["Entitlements"] as? [String: Any],
              let aps = entitlements["aps-environment"] as? String
        else { return "production" }
        // Apple spells it "development"; the APNs host that accepts those tokens is the sandbox one.
        return aps == "development" ? "sandbox" : "production"
    }

    // MARK: - Device token (LeCodesAppHooks forwards, main thread)

    private var deviceTokenHex: String? = nil
    private var tokenWaiters: [(String?) -> Void] = []

    public func remoteNotificationsToken(_ token: Data) {
        let hex = token.map { String(format: "%02x", $0) }.joined()
        onMainThread { [self] in
            let changed = deviceTokenHex != nil && deviceTokenHex != hex
            deviceTokenHex = hex
            let waiters = tokenWaiters
            tokenWaiters = []
            waiters.forEach { $0(hex) }
            // APNs rotated the token (restore, reinstall): every persisted registration
            // points at a dead token — re-register them all with their stored user keys.
            if changed {
                for (uuid, info) in persistedRegistrations() {
                    registerWithBackend(uuid, user: info["user"], token: hex) { _ in }
                }
            }
        }
    }

    public func remoteNotificationsError(_ error: Error) {
        print("LeCodes: APNs registration failed — \(error.localizedDescription)")
        onMainThread { [self] in
            let waiters = tokenWaiters
            tokenWaiters = []
            waiters.forEach { $0(nil) }
        }
    }

    private func withDeviceToken(_ proceed: @escaping (String?) -> Void) {
        if let token = deviceTokenHex { return proceed(token) }
        tokenWaiters.append(proceed)
        UIApplication.shared.registerForRemoteNotifications()
        // APNs normally answers in well under a second; a token that never arrives
        // (no entitlement, no network) must reject, not hang the app's promise.
        DispatchQueue.main.asyncAfter(deadline: .now() + 15) { [self] in
            guard !tokenWaiters.isEmpty else { return }
            let waiters = tokenWaiters
            tokenWaiters = []
            waiters.forEach { $0(deviceTokenHex) }
        }
    }

    // MARK: - Persistence (per-project registrations)

    private let registrationsKey = "lecodes.push.registrations"

    private func persistedRegistrations() -> [String: [String: String]] {
        (UserDefaults.standard.dictionary(forKey: registrationsKey) as? [String: [String: String]]) ?? [:]
    }
    private func persistRegistration(_ uuid: String, address: String, user: String?) {
        var all = persistedRegistrations()
        var info = ["address": address]
        if let user { info["user"] = user }
        all[uuid] = info
        UserDefaults.standard.set(all, forKey: registrationsKey)
    }
    private func removeRegistration(_ uuid: String) {
        var all = persistedRegistrations()
        all.removeValue(forKey: uuid)
        UserDefaults.standard.set(all, forKey: registrationsKey)
    }

    // MARK: - Service calls (invoked by PushService on the main thread)

    func getStatus(_ projectUuid: String?, _ settle: ChannelSettle) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let permission: String
            switch settings.authorizationStatus {
            case .authorized, .provisional, .ephemeral: permission = "granted"
            case .denied: permission = "denied"
            default: permission = "prompt"
            }
            let registered = projectUuid != nil && self.persistedRegistrations()[projectUuid!] != nil
            settle.resolve(["permission": permission, "registered": registered])
        }
    }

    func register(_ projectUuid: String?, _ options: [String: Any]?, _ settle: ChannelSettle) {
        // No identity = an unattributed world (launcher, dev bundle): push can't know which
        // app is asking, so it fails BEFORE any permission path (the plugin ordering rule).
        guard let uuid = projectUuid else { return settle.reject("unavailable") }
        let user = options?["user"] as? String

        // Straight to the OS prompt — no viewer-level consent sheet. Turning a project's
        // notifications OFF is enforced at the RELAY, not here: when the app isn't running iOS
        // displays the push itself and no LeCodes code runs, so a device-side gate could never be
        // the real control. Per-project opt-out lives on the registration row (`Push.unregister()`,
        // and the mute the launcher sets) — see docs/push-plan.md.
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, _ in
            onMainThread { [self] in
                guard granted else { return settle.reject("denied") }
                withDeviceToken { [self] token in
                    guard let token else { return settle.reject("unavailable") }
                    registerWithBackend(uuid, user: user, token: token) { address in
                        if let address {
                            settle.resolve(["address": address])
                        } else {
                            settle.reject("unavailable")
                        }
                    }
                }
            }
        }
    }

    func unregister(_ projectUuid: String?, _ settle: ChannelSettle) {
        guard let uuid = projectUuid else { return settle.reject("unavailable") }
        if let address = persistedRegistrations()[uuid]?["address"], let token = deviceTokenHex {
            var request = URLRequest(url: URL(string: "\(apiBase)/api/push/registrations/\(address)")!)
            request.httpMethod = "DELETE"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try? JSONSerialization.data(withJSONObject: ["token": token])
            URLSession.shared.dataTask(with: request).resume()   // best effort
        }
        removeRegistration(uuid)
        settle.resolve()
    }

    func getLaunch(_ projectUuid: String?, _ settle: ChannelSettle) {
        guard let uuid = projectUuid, let payload = launchPayload,
              payload.projectUuid == uuid else { return settle.resolve() }
        settle.resolve(payload.appPayload)
    }

    // MARK: - Backend registration

    private func registerWithBackend(_ uuid: String, user: String?, token: String,
                                     _ done: @escaping (String?) -> Void) {
        var body: [String: Any] = [
            "projectUuid": uuid, "platform": "apns", "token": token, "environment": environment,
        ]
        // The relay picks send credentials — and the apns-topic — from the host app's bundle
        // id. A standalone shell MUST send its own: registered under the viewer's topic its
        // sends would die with DeviceTokenNotForTopic. The relay also fail-fasts registration
        // for bundle ids it can never sign for (foreign teams, until per-app keys exist).
        if let hostApp = Bundle.main.bundleIdentifier { body["hostApp"] = hostApp }
        if let user { body["user"] = user }
        var request = URLRequest(url: URL(string: "\(apiBase)/api/push/registrations")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        URLSession.shared.dataTask(with: request) { data, response, _ in
            var address: String? = nil
            if let data, (response as? HTTPURLResponse)?.statusCode == 200,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                address = json["address"] as? String
            }
            onMainThread { [self, address] in
                if let address { persistRegistration(uuid, address: address, user: user) }
                done(address)
            }
        }.resume()
    }

    // MARK: - Delivery (LeCodesAppHooks handler + cold start)

    /// One parsed notification. `appPayload` is what JS sees (routing fields stripped).
    private struct Payload {
        let projectUuid: String
        let appPayload: [String: Any]
        let url: String?
    }

    /// The notification that launched (or re-routed) the current world — served by
    /// getLaunch to the matching project's world, stable for that world's lifetime.
    private var launchPayload: Payload? = nil

    private func parse(_ userInfo: [AnyHashable: Any]) -> Payload? {
        guard let lecodes = userInfo["lecodes"] as? [String: Any],
              let uuid = lecodes["projectUuid"] as? String else { return nil }
        var app: [String: Any] = [:]
        if let aps = userInfo["aps"] as? [String: Any] {
            if let alert = aps["alert"] as? [String: Any] {
                app["title"] = alert["title"] ?? ""
                if let body = alert["body"] { app["body"] = body }
            }
            if let badge = aps["badge"] { app["badge"] = badge }
        }
        if let url = lecodes["url"] { app["url"] = url }
        if let data = lecodes["data"] { app["data"] = data }
        return Payload(projectUuid: uuid, appPayload: app, url: lecodes["url"] as? String)
    }

    /// `connectionOptions.notificationResponse` (buffered by LeCodesAppHooks) — the app was
    /// cold-started by a notification tap. The plugin registers BEFORE the JS world boots
    /// (registerBundledPlugins / postInit precede run), so the seeded launch URL is what
    /// initial code observes — the same seeding order rule as deep links.
    public func coldStart(_ response: UNNotificationResponse) {
        guard let engine,
              let payload = parse(response.notification.request.content.userInfo) else { return }
        launchPayload = payload
        coldStartHandledId = response.notification.request.identifier
        engine.setLaunchUrl("https://le.codes/qr/\(payload.projectUuid)")
    }

    /// The notification this launch was started by, once the cold-start path has consumed it.
    ///
    /// iOS reports a launching tap TWICE — through `connectionOptions.notificationResponse` and
    /// again to the delegate as `didReceive`. The second delivery lands while the launcher is still
    /// fetching the project, so `currentProjectUuid` has not been set yet, and `didReceive` reads
    /// that as "a notification for some other app" and re-enters the launcher. The world is then
    /// built twice, and everything the first one had in flight (its `register()` above all) dies
    /// with its context. Same notification, already routed — the redelivery is nothing to act on.
    /// (LeCodesAppHooks drains the cold-start response before any buffered didReceive, so the id
    /// is always set before the redelivery is seen.)
    private var coldStartHandledId: String? = nil

    /// Foreground receipt: the target app is on screen → no banner, deliver as the
    /// "message" event; any other app's notification shows normally.
    public func willPresent(_ notification: UNNotification,
                            completion completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        guard let payload = parse(notification.request.content.userInfo) else {
            return completionHandler([.banner, .sound, .list])
        }
        onMainThread { [self] in
            if payload.projectUuid == engine?.currentProjectUuid {
                emit("message", payload.appPayload)
                completionHandler([])
            } else {
                completionHandler([.banner, .sound, .list])
            }
        }
    }

    /// Tap while the process is alive: same-project world → warm "tap" event (+ the url
    /// riding the deep-link path); any other project → re-enter the launcher with a seeded
    /// QR URL (the same primitive quit() uses) and serve the payload via getLaunch.
    public func didReceive(_ response: UNNotificationResponse,
                           completion completionHandler: @escaping () -> Void) {
        guard let payload = parse(response.notification.request.content.userInfo) else {
            return completionHandler()
        }
        // Already consumed by the cold-start path — see `coldStartHandledId`. Acting on it again
        // would rebuild the world the launcher is in the middle of building.
        if coldStartHandledId == response.notification.request.identifier {
            coldStartHandledId = nil
            return completionHandler()
        }
        onMainThread { [self] in
            guard let engine else { return completionHandler() }
            if payload.projectUuid == engine.currentProjectUuid {
                emit("tap", payload.appPayload)
                if let url = payload.url { engine.emitUrl(url) }
            } else {
                launchPayload = payload
                engine.openLauncher(withUrl: "https://le.codes/qr/\(payload.projectUuid)")
            }
            completionHandler()
        }
    }

    // MARK: - Live sessions (the event pipe)

    private var sessions: [WeakPushService] = []
    private struct WeakPushService { weak var service: PushService? }

    func attach(_ service: PushService) { sessions.append(.init(service: service)) }
    func detach(_ service: PushService) { sessions.removeAll { $0.service === service || $0.service == nil } }

    private func emit(_ event: String, _ payload: [String: Any]) {
        sessions.removeAll { $0.service == nil }
        for session in sessions { session.service?.channel.emit(event, payload) }
    }
}

File: plugins/push/android/src/main/java/io/letary/lecodes/plugins/push/PushManager.kt
//
//  PushManager.kt — lecodes-plugins/push
//
//  The Android mirror of the iOS push plugin's PushManager: FCM token + backend registration
//  + payload routing behind the "push" service (docs/push-plan.md).
//
//  Contract notes carried over from the iOS device pass:
//   - register() checks identity BEFORE the permission prompt ("unavailable" beats the dialog).
//   - getStatus() never prompts and works without identity (the SDK pokes the session open with
//     it to establish the event pipe).
//   - Token rotation mints NEW addresses (the backend upsert key is the token) — onNewToken
//     re-registers every persisted project and stores the fresh address.
//

package io.letary.lecodes.plugins.push

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.google.firebase.messaging.FirebaseMessaging
import io.letary.lecodes.LecodesEngine
import io.letary.lecodes.utils.PermissionRequester
import io.letary.lecodes.utils.Services
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

object PushManager {

    private const val API_BASE = "https://le.codes"
    const val EXTRA_PAYLOAD = "lecodes.push.payload"

    private lateinit var appContext: Context
    var engine: LecodesEngine? = null
        private set

    /** PushPlugin.setForeground (host onResume/onPause) — a banner is suppressed only for a
     *  foregrounded same-project world (the iOS willPresent rule). */
    @Volatile var isForeground = false

    private val sessions = mutableListOf<Services.Channel>()
    private var launchPayload: JSONObject? = null

    // The two backend calls are tiny JSON requests; a dedicated thread keeps them off the main
    // thread without touching the engine's shared OkHttp client (which may not exist yet when a
    // background FCM start reaches this object).
    private val httpExecutor = Executors.newSingleThreadExecutor()

    fun init(context: Context, engine: LecodesEngine) {
        appContext = context.applicationContext
        this.engine = engine
    }

    /** Bootstrap path for FCM service entry points that can run before the host activity did. */
    fun ensureContext(context: Context) {
        if (!this::appContext.isInitialized) appContext = context.applicationContext
    }

    private val prefs: SharedPreferences
        get() = appContext.getSharedPreferences("lecodes.push", Context.MODE_PRIVATE)

    // ── sessions (the "push" service's event pipe) ───────────────────────────

    fun addSession(channel: Services.Channel) = synchronized(sessions) { sessions.add(channel) }
    fun removeSession(channel: Services.Channel) = synchronized(sessions) { sessions.remove(channel) }

    private fun emitToSessions(event: String, payload: JSONObject) {
        val snapshot = synchronized(sessions) { sessions.toList() }
        for (channel in snapshot) channel.emit(event, payload.toString())
    }

    // ── service calls ────────────────────────────────────────────────────────

    fun getStatus(): JSONObject {
        val permission = when {
            Build.VERSION.SDK_INT < 33 ->
                if (notificationsEnabled()) "granted" else "denied"
            ContextCompat.checkSelfPermission(appContext, Manifest.permission.POST_NOTIFICATIONS) ==
                PackageManager.PERMISSION_GRANTED -> if (notificationsEnabled()) "granted" else "denied"
            prefs.getBoolean("permissionRequested", false) -> "denied"
            else -> "prompt"
        }
        val registered = engine?.currentProjectUuid?.let { loadRegistration(it) != null } ?: false
        return JSONObject().put("permission", permission).put("registered", registered)
    }

    fun register(call: Services.ServiceCall, user: String?) {
        // Identity FIRST — an unattributed world must reject before any permission path.
        val uuid = engine?.currentProjectUuid ?: return call.reject("unavailable")
        withPermission(call) {
            FirebaseMessaging.getInstance().token
                .addOnSuccessListener { token ->
                    postRegistration(uuid, user, token) { address, error ->
                        if (address != null) {
                            saveRegistration(uuid, address, user)
                            call.resolve(JSONObject().put("address", address))
                        } else {
                            call.reject(error ?: "unavailable")
                        }
                    }
                }
                .addOnFailureListener { call.reject("unavailable") }
        }
    }

    fun unregister(call: Services.ServiceCall) {
        val uuid = engine?.currentProjectUuid ?: return call.reject("unavailable")
        val entry = loadRegistration(uuid)
        if (entry == null) { call.resolve(); return }
        removeRegistration(uuid)
        // Fire-and-forget like iOS — the raw token is the ownership proof the DELETE carries.
        FirebaseMessaging.getInstance().token.addOnSuccessListener { token ->
            httpExecutor.execute {
                runCatching {
                    httpJson("DELETE", "$API_BASE/api/push/registrations/${entry.optString("address")}",
                        JSONObject().put("token", token))
                }
            }
        }
        call.resolve()
    }

    /** The notification that cold-started this world — gated on the CURRENT world's identity so
     *  another project's parked payload never leaks into a different app. */
    fun getLaunch(): JSONObject? {
        val payload = launchPayload ?: return null
        if (payload.optString("projectUuid") != engine?.currentProjectUuid) return null
        return toJsPayload(payload)
    }

    // ── FCM entry points (LePushMessagingService) ────────────────────────────

    /** @return true when handled as a foreground same-project "message" (banner suppressed). */
    fun onMessage(payload: JSONObject): Boolean {
        val e = engine ?: return false
        val uuid = e.currentProjectUuid ?: return false
        if (!isForeground || uuid != payload.optString("projectUuid")) return false
        emitToSessions("message", toJsPayload(payload))
        return true
    }

    fun onNewToken(token: String) {
        val regs = loadRegistrations()
        for (uuid in regs.keys()) {
            val entry = regs.optJSONObject(uuid) ?: continue
            val user = entry.optString("user").takeIf { it.isNotEmpty() }
            // Rotation mints a new address (backend upsert key is the token) — store the fresh one.
            postRegistration(uuid, user, token) { address, _ ->
                if (address != null) saveRegistration(uuid, address, user)
            }
        }
    }

    // ── tap routing (PushPlugin → host activity) ─────────────────────────────

    /** Cold start: park the payload for getLaunch() BEFORE the engine boots. The intent's
     *  le.codes/qr data URI rides the normal launchUrl path unchanged. The extra is consumed so
     *  an Activity recreation doesn't re-park it (the iOS coldStartHandledId analog). */
    fun parkLaunchPayload(intent: Intent) {
        val raw = intent.getStringExtra(EXTRA_PAYLOAD) ?: return
        intent.removeExtra(EXTRA_PAYLOAD)
        launchPayload = runCatching { JSONObject(raw) }.getOrNull()
    }

    /** Warm tap (onNewIntent). @return true when consumed — the caller must then SKIP the
     *  generic deep-link path (a same-project tap must not reload the running app). */
    fun handleTapIntent(intent: Intent): Boolean {
        val raw = intent.getStringExtra(EXTRA_PAYLOAD) ?: return false
        intent.removeExtra(EXTRA_PAYLOAD)
        val payload = runCatching { JSONObject(raw) }.getOrNull() ?: return true
        val e = engine ?: return true
        val projectUuid = payload.optString("projectUuid")
        if (projectUuid == e.currentProjectUuid) {
            emitToSessions("tap", toJsPayload(payload))
            val url = payload.optString("url")
            if (url.isNotEmpty()) e.emitUrl(url)
        } else {
            launchPayload = payload   // the new world reads it via getLaunch()
            e.openLauncher("https://le.codes/qr/$projectUuid")
        }
        return true
    }

    // ── payload shapes ───────────────────────────────────────────────────────

    /** The FCM data map (all strings, buildFcmBody's shape) → internal payload. null = not ours. */
    fun payloadFromData(data: Map<String, String>): JSONObject? {
        val uuid = data["projectUuid"] ?: return null
        val payload = JSONObject()
            .put("projectUuid", uuid)
            .put("title", data["title"] ?: "")
        data["projectName"]?.let { payload.put("projectName", it) }
        data["body"]?.let { payload.put("body", it) }
        data["url"]?.let { payload.put("url", it) }
        data["badge"]?.toIntOrNull()?.let { payload.put("badge", it) }
        data["data"]?.let { raw -> runCatching { payload.put("data", JSONObject(raw)) } }
        return payload
    }

    /** Strip the routing fields before anything reaches JS (the host-side contract). */
    private fun toJsPayload(payload: JSONObject): JSONObject {
        val js = JSONObject()
        js.put("title", payload.optString("title"))
        if (payload.has("body")) js.put("body", payload.optString("body"))
        if (payload.has("url")) js.put("url", payload.optString("url"))
        if (payload.has("badge")) js.put("badge", payload.optInt("badge"))
        payload.optJSONObject("data")?.let { js.put("data", it) }
        return js
    }

    // ── permission (house rule: prompt on first call, never in factory) ──────

    private fun notificationsEnabled() = NotificationManagerCompat.from(appContext).areNotificationsEnabled()

    private fun withPermission(call: Services.ServiceCall, body: () -> Unit) {
        if (Build.VERSION.SDK_INT < 33) {
            if (notificationsEnabled()) body() else call.reject("denied")
            return
        }
        if (ContextCompat.checkSelfPermission(appContext, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED) return body()
        val e = engine ?: return call.reject("denied")
        if (e.getActivity() == null) return call.reject("denied")
        prefs.edit().putBoolean("permissionRequested", true).apply()
        PermissionRequester.request(e, arrayOf(Manifest.permission.POST_NOTIFICATIONS)) { granted ->
            if (granted) body() else call.reject("denied")
        }
    }

    // ── persistence (the iOS UserDefaults shape: projectUuid → { address, user }) ────────────

    private fun loadRegistrations(): JSONObject =
        runCatching { JSONObject(prefs.getString("registrations", null) ?: "{}") }.getOrDefault(JSONObject())

    private fun loadRegistration(projectUuid: String): JSONObject? =
        loadRegistrations().optJSONObject(projectUuid)

    private fun saveRegistration(projectUuid: String, address: String, user: String?) {
        val regs = loadRegistrations()
        val entry = JSONObject().put("address", address)
        if (user != null) entry.put("user", user)
        regs.put(projectUuid, entry)
        prefs.edit().putString("registrations", regs.toString()).apply()
    }

    private fun removeRegistration(projectUuid: String) {
        val regs = loadRegistrations()
        regs.remove(projectUuid)
        prefs.edit().putString("registrations", regs.toString()).apply()
    }

    // ── backend ──────────────────────────────────────────────────────────────

    private fun postRegistration(
        projectUuid: String, user: String?, token: String,
        onResult: (address: String?, error: String?) -> Unit,
    ) {
        val body = JSONObject()
            .put("projectUuid", projectUuid)
            .put("platform", "fcm")
            .put("token", token)
            .put("hostApp", appContext.packageName)
        if (user != null) body.put("user", user)

        httpExecutor.execute {
            val response = runCatching { httpJson("POST", "$API_BASE/api/push/registrations", body) }.getOrNull()
            when {
                response == null -> onResult(null, "unavailable")
                response.first in 200..299 -> {
                    val address = runCatching { JSONObject(response.second).optString("address") }
                        .getOrNull()?.takeIf { it.isNotEmpty() }
                    onResult(address, if (address == null) "registration failed" else null)
                }
                else -> onResult(null, runCatching { JSONObject(response.second).optString("error") }
                    .getOrNull()?.takeIf { it.isNotEmpty() } ?: "registration failed (${response.first})")
            }
        }
    }

    private fun httpJson(method: String, url: String, body: JSONObject): Pair<Int, String> {
        val connection = URL(url).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.connectTimeout = 10_000
            connection.readTimeout = 10_000
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { it.write(body.toString().toByteArray()) }
            val code = connection.responseCode
            val text = (if (code in 200..299) connection.inputStream else connection.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: ""
            return code to text
        } finally {
            connection.disconnect()
        }
    }
}
